File: AMFICOM/v1/dadara/native/analyse/Histogramm.cpp
// Histogramm.cpp: implementation of the Histogramm class.
//
//////////////////////////////////////////////////////////////////////

#include "Histogramm.h"

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////

Histogramm::Histogramm(double down_limit, double up_limit, int nBins)
{
	if (down_limit > up_limit)
	{
		double tmp = down_limit;
		down_limit = up_limit;
		up_limit = tmp;
	}

	this->up_limit = up_limit;
	this->down_limit = down_limit;
	this->nBins	= nBins;
}

Histogramm::~Histogramm()
{
	delete histo;
}
 
void Histogramm::init(double* data, int data_length, int start, int end)
{
	histo = new double[nBins];

	for (int i = 0; i < nBins; i++)
		histo[i] = 0;

	double deriv_delta = (up_limit - down_limit) / (double)nBins;

	int N;
	for (i = max(0, start); i <= min (end, data_length-1); i++)
	{
		N = (int)((data[i] - down_limit) / deriv_delta + 0.5);
		if (N >= 0 && N < nBins)
			histo[N]++;
	}
}

int Histogramm::getMaximumIndex()
{
	int center = 0;
	double max = histo[0];
	for (int i = 1; i < nBins; i++)
		if (max < histo[i])
		{
			max = histo[i];
			center = i;
		}
	return center;
}

double Histogramm::getMaximumValue()
{
	double deriv_delta = (up_limit - down_limit) / (double)nBins;
	int max_index = getMaximumIndex();
	return down_limit + (max_index + .5) * deriv_delta;
}

File: AMFICOM/v1/analysis/src/com/syrus/AMFICOM/analysis/dadara/Histogramm.java
package com.syrus.AMFICOM.analysis.dadara;

public class Histogramm {
	private double upLimit;
	private double downLimit;
	private int nBins;
	private double[] histo;

	public Histogramm(double downLimit, double upLimit, int nBins) {
		if (downLimit > upLimit) {
			double tmp = downLimit;
			downLimit = upLimit;
			upLimit = tmp;
		}
		this.upLimit = upLimit;
		this.downLimit = downLimit;
		this.nBins = nBins;
	}

	public double[] init(double[] data, int start, int end) {
		this.histo = new double[this.nBins];
		double derivDelta = (this.upLimit - this.downLimit) / this.nBins;

		int n;
		for (int i = Math.max(0, start); i <= Math.min(end, data.length - 1); i++) {
			n = (int) Math.round((data[i] - this.downLimit) / derivDelta);
			if (n >= 0 && n < this.nBins) {
				this.histo[n]++;
			}
		}
		return this.histo;
	}

	public int getMaximumIndex() {
		int center = 0;
		double max = this.histo[0];
		for (int i = 1; i < this.histo.length; i++) {
			if (this.histo[i] > max) {
				max = this.histo[i];
				center = i;
			}
		}
		return center;
	}

	public double getMaximumValue() {
		int max = getMaximumIndex();
		return this.downLimit + (this.upLimit - this.downLimit) / this.nBins * max;
	}

	public double[] getHistogramm() {
		return this.histo;
	}
}
